File: src/main/GO/leetcode/lists/hot100/LeetCode152.go
package hot100

func maxProduct(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	max := nums[0]
	min := nums[0]
	res := nums[0]
	for i := 1; i < len(nums); i++ {
		mx, mn := max, min
		max = Max(Max(mx*nums[i], mn*nums[i]), nums[i])
		min = Min(Min(mx*nums[i], mn*nums[i]), nums[i])
		res = Max(res, max)
	}
	return res
}

File: src/main/java/leetcode/lists/hot100/LeetCode152.java
package leetcode.lists.hot100;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/4 17:03
 */
public class LeetCode152 {

    //https://www.youtube.com/watch?v=0Kpz-ChuQIE

    public int maxProduct(int[] nums) {

        if (nums == null || nums.length == 0) {
            return 0;
        }
        int maxF = nums[0], minF = nums[0], ans = nums[0];

        for (int i = 1; i < nums.length; i++) {
            int mx = maxF, mn = minF;
            maxF = Math.max(nums[i], Math.max(mx * nums[i], mn * nums[i]));
            minF = Math.min(nums[i], Math.min(mx * nums[i], mn * nums[i]));
            ans = Math.max(ans, maxF);
        }
        return ans;
    }
}
